knimin/static/js/platemapper.js

```diff
@@ -0,0 +1,506 @@
+    // information
+    this.samples = {}
+    // This is a special list needed for initializing the sample autocompletion
+    this.autoCompleteSamples = []
+    // Iterate over all the studies
+    for (var idx = 0; idx < this.studies.length; idx++) {
+      study = this.studies[idx];
+      color = PlateMap._qiimeColors[idx];
+      // Iterate over all samples
+      for (var sample of study.samples.all) {
+        this.samples[sample] = {color: color, plates: []};
+        this.autoCompleteSamples.push({label: sample, category: study.title, color: color});
+      }
+      // Iterate over all plates to get the already plated samples
+      for (var plate in study.samples.plated) {
+        if (study.samples.plated.hasOwnProperty(plate)) {
+          this.samples[sample].plates.push(plate);
+        }
+      }
+    }
+
+    // Create a 2D array to store the per well information
+    this.wellInformation = new Array(this.rows);
+    for (var i = 0; i < this.wellInformation.length; i++) {
+      this.wellInformation[i] = new Array(this.cols);
+      for (var j = 0; j < this.wellInformation[i].length; j++) {
+        this.wellInformation[i][j] = {inputTag: undefined, comment: undefined};
+      }
+    }
+
+    this.drawPlate();
+};
+
+/**
+ *
+ * Updates the contents of the text area with a summary of the well comments
+ *
+ **/
+PlateMap.prototype.updateWellCommentsArea = function () {
+  var sample, wellId, wellInfo;
+  var comments = "";
+  for (var i = 0; i < this.rows; i++) {
+    for (var j = 0; j < this.cols; j++) {
+      wellInfo = this.wellInformation[i][j];
+      if (wellInfo.comment !== undefined) {
+        sample = wellInfo.inputTag.val().trim();
+        if (sample.length === 0) {
+          sample = 'No sample plated';
+        }
+        wellId = String.fromCharCode('A'.charCodeAt() + i) + (j + 1);
+        comments = comments + "Well " + wellId + " (Sample: " + sample + "): " + wellInfo.comment + "\n";
+      }
+    }
+  }
+  $("#well-comments-area").val(comments);
+}
+
+/**
+ *
+ * Constructs the layout to be sent to the server
+ *
+ * @return {Array}
+ *
+ **/
+PlateMap.prototype._construct_layout = function () {
+  var layout, row, wellInfo;
+
+  layout = []
+  for (var wellRow of this.wellInformation) {
+    row = []
+    for (var well of wellRow) {
+      wellInfo = {};
+      wellInfo.notes = well.comment;
+      sample = well.inputTag.val().trim();
+      if (sample.length > 0) {
+        if (sample in this.samples) {
+          wellInfo.sample_id = sample;
+        } else {
+          wellInfo.name = sample;
+        }
+      }
+      row.push(wellInfo);
+    }
+    layout.push(row);
+  }
+  return layout;
+}
+
+/**
+ *
+ * Saves the current layout in the server
+ *
+ **/
+PlateMap.prototype.savePlate = function () {
+
+  var layout = this._construct_layout();
+  $.post($(location).attr('href'), {action: 'save', layout: JSON.stringify(layout), plate_id: this.plateId}, function(data) {
+    console.log(data);
```

Comment:
Instead of keeping this private to the developer, maybe showing a small indicator on the interface that says "data has been saved to the server" would be reassuring and useful to the user?